session.go

```diff
@@ -226,27 +226,107 @@ func (s *Session) init() error {
-
-		hosts = append(hosts, host)
-	}
-
 	type bulkAddHosts interface {
 		AddHosts([]*HostInfo)
 	}
-	if v, ok := s.policy.(bulkAddHosts); ok {
-		v.AddHosts(hosts)
+	bulkAddHostsPolicy, _ := s.policy.(bulkAddHosts)
+
+	// if the policy supports Ready() then we'll call it until it's ready and then return
+	if rdy, ok := s.policy.(ReadyPolicy); ok {
+		// each host will increment left and decrement it after connecting and once
+		// there's none left, we'll close hostCh
+		var left int64
+		// we will receive up to len(hostMap) of messages so create a buffer so we
+		// don't end up stuck in a goroutine if we stopped listening
+		hostCh := make(chan *HostInfo, len(hostMap))
+		// we add one here because we don't want to end up closing hostCh until we're
+		// done looping and the decerement code might be reached before we've looped
+		// again
+		atomic.AddInt64(&left, 1)
+		for _, h := range hostMap {
+			host := s.ring.addOrUpdate(h)
+			if s.cfg.filterHost(host) {
+				continue
+			}
+
+			atomic.AddInt64(&left, 1)
+			go func() {
+				host.setState(NodeUp)
+				s.pool.addHost(host)
+				hostCh <- host
+
+				// if there are no hosts left, then close the hostCh to unblock the loop
+				// below if its still waiting
+				if atomic.AddInt64(&left, -1) == 0 {
+					close(hostCh)
+				}
+			}()
+		}
+		// once we're done looping we subtract the one we initially added and check
+		// to see if we should close
+		if atomic.AddInt64(&left, -1) == 0 {
+			close(hostCh)
+		}
+		// if the policy supports bulk adding, then we should try to read as much as
+		// possible at a given time (within 10ms) in order to add as many hosts at
+		// once as we can
+		if bulkAddHostsPolicy != nil {
+			for {
+				host, ok := <-hostCh
+				if !ok {
+					break
+				}
+				// after we've read a single host, wait 10ms to collect any other ones
+				hosts := []*HostInfo{host}
+			Outer:
+				for {
+					after := time.After(10 * time.Millisecond)
+					select {
+					case host, ok := <-hostCh:
+						// if hostCh is closed, immediately break
+						if !ok {
+							break Outer
+						}
+						hosts = append(hosts, host)
+					case <-after:
+						break Outer
+					}
+				}
+				bulkAddHostsPolicy.AddHosts(hosts)
+				if rdy.Ready() {
+					break
+				}
+			}
+		} else {
+			for host := range hostCh {
+				s.policy.AddHost(host)
+				if rdy.Ready() {
+					break
+				}
+			}
+		}
 	} else {
-		for _, host := range hosts {
-			s.policy.AddHost(host)
+		// otherwise since we don't care about readiness, we just connect to all
+		// hosts and then add them to the policy afterwards
+		hosts = hosts[:0]
+		for _, host := range hostMap {
+			host = s.ring.addOrUpdate(host)
+			if s.cfg.filterHost(host) {
+				continue
+			}
+
+			host.setState(NodeUp)
+			s.pool.addHost(host)
+
+			hosts = append(hosts, host)
+		}
+
+		if bulkAddHostsPolicy != nil {
+			bulkAddHostsPolicy.AddHosts(hosts)
+		} else {
+			for _, host := range hosts {
+				s.policy.AddHost(host)
+			}
```

Comment:
This branch still connects to the hosts serially. I think even in the case when we want to wait for all hosts to be connected we want to be connecting to them concurrently, so we could probably remove this branch altogether.

Reply:
👍 can do!